ggml-cuda.cu

```diff
@@ -5987,6 +5987,88 @@ static  __global__ void im2col_f32_f16(
     }
 }
 
+#define CUDA_FLASH_ATTENTION_BLOCK_SIZE 256
+
+template<int block_size>
+static __global__ void flash_attn_f32(const float* q, const float* k,const float* v, float* dst, float kq_scale,
+    int d_head, int seq_len, int num_heads) {
+    const int head = blockIdx.x / seq_len;
+    const int head_size = d_head * seq_len;
+    const int s = blockIdx.x % seq_len;
+    const int tid = threadIdx.x;
+
+    extern __shared__  char work_data[];
+    float* S = (float*)work_data; // theorical sequent length: 12848, due memory per block limit
+    float* warp_data = (float*)(work_data + seq_len * sizeof(float));
```

Comment:
Tensor cores can handle FP16, BF16, and TF32 but **not** regular floats. So computing data as FP32 is fine for a test implementation to ensure that results are correct but for the actual production implementation you will need to use something else.

Also I don't understand what that comment is supposed to tell me.

Reply:
At the moment, it is set up like this. I know that currently the best approach would be to use tensor cores for performing GEMM in this kernel. However, I wonder if I should then convert from f32 to f16 before executing the flash attention kernel and utilize tensor cores to accelerate GEMM.